NDTensors/src/lib/Sectors/test/test_category_product.jl

```diff
@@ -60,10 +60,10 @@ using Test: @test, @testset, @test_throws
     q21 = (N=U1(2),) × (J=SU2(1),)
     q22 = (N=U1(2),) × (J=SU2(2),)
 
-    @test q1h ⊗ q1h == q20 ⊕ q21
+    @test q1h ⊗ q1h == gradedrange([q20 => 1, q21 => 1])
```

Comment:
So I realized an issue with tests like this is that `GradedUnitRange` is just use a generic range comparison, so ignores the labels in the comparison.

It may be best to write these as:
```julia
@test blocklengths(q1h ⊗ q1h) == [1, 1]
@test blocklabels(q1h ⊗ q1h) == [q20, q21]
```

Reply:
I agree that having `AbstractCategory` acting as a `GradedUnitRange` would be convenient. However I am annoyed by the implications. Does it mean there is no simple way to check whether 2 `GradedUnitRange` are equal? That would be an issue for `FusionTensor`.

Reply:
We can add a separate function to check equality of graded axes, we just can't use `==` for that. For `GradedUnitRange` to act like other ranges (i.e. adhere to the `AbstractUnitRange` interface), equality is based on iterating through the values of the range.

An analogy is that in `BlockArrays`, blocked ranges with different block structures are considered equal, since they iterate over the same values, and there is another interface [blockisequal](https://juliaarrays.github.io/BlockArrays.jl/stable/lib/public/#BlockArrays.blockisequal) for additionally checking if the block structure is the same:
```julia
julia> using BlockArrays

julia> b1 = blockedrange([1, 2])
2-blocked 3-element BlockedUnitRange{Vector{Int64}}:
 1
 ─
 2
 3

julia> b2 = blockedrange([1, 1, 1])
3-blocked 3-element BlockedUnitRange{Vector{Int64}}:
 1
 ─
 2
 ─
 3

julia> b1 == 1:3
true

julia> b2 == 1:3
true

julia> blockisequal(b1, b2)
false

julia> blockisequal(b1, 1:3)
false

julia> blockisequal(b2, 1:3)
false
```
By analogy, we could define `gradedisequal` as:
```julia
function gradedisequal(a1::AbstractUnitRange, a2::AbstractUnitRange)
  return blockisequal(a1, a2) && (blocklabels(a1) == blocklabels(a2))
end
```
I realize that seeing these kind of unintuitive interface decisions can be frustrating, however a lot of it is constrained by the choice of wanting `GradedUnitRange` to be an `AbstractUnitRange`, so that it can be directly used as axes of an `AbstractArray`. I think that choice will pay off, but it comes at the expense of being constrained by that interface.

Reply:
So with that suggestion, you could simply rewrite this kind of test as:
```julia
    @test gradedisequal(q1h ⊗ q1h, gradedrange([q20 => 1, q21 => 1]))
```

Reply:
A slightly kooky idea is that we can define:
```julia
==ᵍ(x, y) = gradedisequal(x, y)
```
(written `==\^g<tab>` in the REPL or in an IDE with a Julia extension: https://github.com/JuliaEditorSupport). Then we can use:
```julia
    @test q1h ⊗ q1h ==ᵍ gradedrange([q20 => 1, q21 => 1]))
```
We could be conservative about it and only use that for convenience in the tests.

Also, as more context for why the current `==` definition for `GradedUnitRange` makes sense, it is comparable how `==` works (and should work) for other kinds of arrays, where array equality is based on whether the elements of the array are equal, independent of the structure of the array. For example:
```julia
julia> using SparseArrays

julia> using BlockArrays

julia> S = sprand(4, 4, 0.5)
4×4 SparseMatrixCSC{Float64, Int64} with 9 stored entries:
 0.466251  0.162259   ⋅         ⋅ 
 0.821829  0.531673   ⋅        0.397923
 0.170507   ⋅        0.207483  0.552266
 0.692108   ⋅         ⋅         ⋅ 

julia> A = Array(S)
4×4 Matrix{Float64}:
 0.466251  0.162259  0.0       0.0
 0.821829  0.531673  0.0       0.397923
 0.170507  0.0       0.207483  0.552266
 0.692108  0.0       0.0       0.0

julia> B = BlockArray(S, [2, 2], [2, 2])
2×2-blocked 4×4 BlockMatrix{Float64, Matrix{SparseMatrixCSC{Float64, Int64}}, Tuple{BlockedUnitRange{Vector{Int64}}, BlockedUnitRange{Vector{Int64}}}}:
 0.466251  0.162259  │   ⋅         ⋅      
 0.821829  0.531673  │   ⋅        0.397923
 ────────────────────┼────────────────────
 0.170507   ⋅        │  0.207483  0.552266
 0.692108   ⋅        │   ⋅         ⋅      

julia> S == A == B
true
```
`AbstractUnitRange` is a subtype of `Array`, so it shouldn't act any different from that. The elements of `r::AbstractUnitRange` are simply the values in the range from `first(r):last(r)`, which is the same whether it is a `UnitRange`, `BlockedUnitRange`, or `GradedUnitRange`.

Reply:
@mtfishman just as a question to fill a gap in my understanding, if we defined `==(g1::GradedUnitRange, g2::GradedUnitRange)` specially for only the case of comparing two `GradedUnitRange` types (to have it check labels also), while leaving all other comparisons (say between `GradedUnitRange` and other `AbstractRange` objects) to just check the iterated values, would it break some of the generic capabilities, like the sparse array comparisons above which are quite nice?

Reply:
@ogauthe what do you propose about the `length`, versus `blocklength`, versus `quantum_dimension` interface? At a minimum, we will certainly need some way to distinguish between the notions of "how large is the degeneracy" i.e. size of blocks we store in memory, versus "how big is the vector space" i.e. how big the tensor would be if it became dense. Is your concern more about the  names chosen or that there are three functions versus two? 

E.g. I could see having `nblocks` as an alias of `length` of a `GradedUnitRange` to make it more clear in code what `length` means while still leaving `length` defined to enable generic range code to function.

Reply:
(Also I really like the name of the function `blocklengths` since the "s" at the end makes it clear it will return an array.)

Reply:
It wouldn't make sense to define a special `==` just for `GradedUnitRange`, I'm sure there is generic code within the `AbstractArray` interface, the `AbstractUnitRange` interface, or the `AbstractBlockArray` interface that relies on the current definition, since the current definition is how all `AbstractUnitRange`/axes are defined, though I don't know exactly what the implications would be/what would break. It would make some basic things inconsistent, such as that `==` should be based on comparing the equality of each item getting iterated over, and iteration of ranges is over the range of values, so even that basic assumption would break.

`length(r::GradedUnitRange)` is not equal to `nblocks` for `GradedUnitRange`, it is the length of the range of total values it runs over, i.e. `length(r::GradedUnitRange) = sum(blocklengths(r))`. That is following the definition that `GradedUnitRange` is an `AbstractUnitRange`.

What you are calling `nblocks` is called `blocklength` in `BlockArrays`, i.e.:
```julia
julia> using BlockArrays

julia> blockedrange([2, 2, 2])
3-blocked 6-element BlockedUnitRange{Vector{Int64}}:
 1
 2
 ─
 3
 4
 ─
 5
 6

julia> blocklength(blockedrange([2, 2, 2]))
3
```
which is the same interface I'm using now for `GradedUnitRange`. Personally I also prefer `nblocks` over `blocklength` and maybe will make an alias for that, but for simplicity I'm just sticking to the established `BlockArrays` interface, since at a minimum it makes the type work with the nice functionality in that package for working with blocked ranges and arrays. `blocklength` comes from a convention throughout that package that for a function `f(a::AbstractArray)` that work on a non-blocked array, when it makes sense there is a corresponding function `blockf(a::AbstractArray)` that implements the same function/concept but on the block structure of `a`, so there is `blocksize`, `blocklength`, `blockaxes`, etc. Though in isolation some names end up a bit strange/sub-optimal, I think over all it is a nice naming convention that creates a consistency throughout the package and makes it "obvious" what to name certain functions, so there are less choices one has to make from function to function. Also many functions are defined as `blocksize(a::AbstractArray) = size(blocks(a))`, where `blocks(a::AbstractArray)` for a blocked array is a representation of the blocked array as an array of arrays where the elements of the outer array are the blocks, which I think is a very nice design and convention to follow.

Reply:
@emstoudenmire concerning `length`, I do not think there is an ideal solution in the current design. From a group theory or fusion category perspective, `length` is not really relevant as different sectors live separate lives, what matter is the multiplicity of each block.

I will go on on the basis of a `BlockedUnitRange` and complete fusion rules. Then I will merge `main` into `FusionTensor` and see how things goes.

Reply:
My understanding was that `length` here is just a method overload that enables `BlockedUnitRange` to implement the `AbstractUnitRange` interface and allow iteration and other `AbstractUnitRange` capabilities. So then `length` would not need to really correspond to a group theory concept, but rather more of a range concept.